ref(preprod): Show detailed error for mismatched app comparisons (EME-700)

runningcode opened the pull request:
Fixes https://linear.app/getsentry/issue/EME-700

## Problem

Previous error message when comparing builds that cannot be compared:

```
{"detail": "Head and base size metrics cannot be compared."}
```

## Solution

Enhanced the `can_compare_size_metrics()` function to return detailed error messages that explain exactly why comparison failed:

**Length mismatch example:**
```
"Head and base have different numbers of size metrics. Head has 2 metric(s), base has 1 metric(s)."
```

**Type/identifier mismatch example (Release vs Debug):**
This one might be a bit confusing so I'm open to ways to improve it.
```
"Head and base size metrics cannot be compared due to mismatched metrics. Head has metric(s) not in base: MAIN_ARTIFACT: release. Base has metric(s) not in head: MAIN_ARTIFACT: debug."
```

runningcode committed:
ref(preprod): Show detailed error for mismatched app comparisons (EME-700)

When size metrics cannot be compared, return detailed error messages
explaining exactly why (e.g., different counts, mismatched identifiers
like Release vs Debug). This helps customers understand and fix
comparison failures instead of seeing a generic 400 error.

Changes:
- Enhanced can_compare_size_metrics to return ComparisonValidationResult
- Updated API endpoint to return detailed error messages to users
- Added detailed error logging in background tasks
- Added test for mismatched metric identifiers scenario

runningcode commented on src/sentry/preprod/size_analysis/utils.py:
```diff
@@ -27,18 +36,48 @@ def can_compare_size_metrics(
+
+    if head_only:
+        formatted_head = [format_metric_key(k) for k in head_only]
+        error_parts.append(f"Head has metric(s) not in base: {', '.join(formatted_head)}")
```
This isn't the most user-friendly message. Open to ideas on how to improve it.

trevor-e replied:
Yea I don't think users will understand how we've structured our data as "size metrics". Maybe for this case focus on the app_id? E.g. "`com.hackernews` not found in base build"

runningcode replied:
Yeah, that's a good point. So I added specific messages for the more specific cases but if none of those work then we'll use this generic messages.

trevor-e commented on src/sentry/preprod/size_analysis/utils.py:
```diff
@@ -1,12 +1,21 @@
 
 
+@dataclass
+class ComparisonValidationResult:
```
Off the top of my head I think we have 1) different size metrics, 2) version mismatch, 3) build_configuration mismatch as possible causes to not compare

runningcode replied:
that sounds good. version mismatch shouldn't lead to a comparison error. users want to compare the size across different app versions. or did you mean something else?

trevor-e replied:
analysis version mismatch: https://github.com/getsentry/launchpad/blob/44f628b93167e896752a264eb7d8c7b96ac9bd53/src/launchpad/size/models/common.py#L18-L19

basically if our size analysis code changes substantially the diff might become nonsensical, so we have some semvers to prevent a detailed diff from showing

runningcode committed:
ref(preprod): Add error types for size metric comparison failures (EME-700)

Add ErrorType enum to ComparisonValidationResult to provide stable error
types for frontend error handling. Distinguishes between different length,
different app IDs, different build configurations, and complex mismatches.
Also removes redundant comments and adds comprehensive test coverage.

runningcode committed:
fix(preprod): Fix mypy errors in test_utils (EME-700)

Add assertions that error_message is not None before using in operator
to satisfy mypy type checking.

runningcode committed:
ref(preprod): Restore original comments that weren't added in this PR (EME-700)

Restore comments that existed before our changes to maintain context.
The new error type classification comments added in this PR are kept.

runningcode committed:
fix(preprod): Rename test_utils.py to avoid import conflict (EME-700)

Rename test_utils.py to test_comparison_utils.py to avoid import
conflicts with tests/sentry/profiles/test_utils.py. Pytest was failing
because multiple test files with the same name confuse Python's import
system.
